closure/goog/structs/avltree.js

```diff
@@ -898,4 +899,47 @@ Node.prototype.fixHeight = function() {
       1;
 };
 
+/**
+ * Returns a deep copy of the tree in O(n * k),
+ * where k is the time complexity to copy a node.
+ * @return {AvlTree}
+ */
+AvlTree.prototype.copy = function() {
+  const tree = new AvlTree();
+
+  // Copy instance properties
+  const {copy, leftMost, rightMost} = this.root_.copy();
+  tree.root_       = copy;
+  tree.comparator_ = this.comparator_;
+  tree.minNode_    = leftMost;
+  tree.maxNode_    = rightMost;
+
+  return tree;
+};
+
+Node.prototype.copy = function(parent) {
+  const val  = JSON.parse(JSON.stringify(this.value)); // deep copy
```

Comment:
This only works for basic objects (doesn't copy methods) and will wipe out the prototype chain.

```
class A {
  foo() { console.log('foo'); }
}
const copy = JSON.parse(JSON.stringify(new A()));
copy instanceof A; // false
copy.foo(); // error
```

Maybe have `copy` take an optional copy function? It can use JSON as the default.

```
class A {
  copy() {
    return new A();
  }
}
tree.copy((a) => a.copy());
```

Reply:
By default, `JSON.stringify` will copy the value using the value's `toJSON` method.
Hence, no need to provide an optional copy function.
However, the problem lies in the deserialization.
As u pointed out this way of deep copying the object wipes out the prototype chain.
Hence, i suggest to either take an optional parsing function, or take both an optional copy function and an optional parsing function. What do u prefer?

Reply:
Indeed!
I will add this.

Reply:
I introduced an error in my latest commit.
I don't have the time now but i will fix this ASAP.